### PR TITLE
remove MigrationObject and other swift dependencies from objc code

### PR DESCRIPTION
**Pull request by alazier**
@segiddins @tgoyne @jpsim

### CODE REVIEW

File: RealmSwift/Object.swift [hunk 1]
```diff
@@ -206,7 +216,40 @@ public func == <T: Object>(lhs: T, rhs: T) -> Bool {
     return RLMObjectBaseAreEqual(lhs, rhs)
 }
 
-/// Internal class. Do not use directly.
+/// Object interface which allows untyped getters and setters for Objects.
+public final class DynamicObject : Object {
+    private var listProperties = [String: List<DynamicObject>]()
+
+    /// Returns the value of the property with the given name.
```
Comment (jpsim) [inline RealmSwift/Object.swift]: > Returns or sets

File: RealmSwift/Object.swift [hunk 2]
```diff
@@ -206,7 +216,40 @@ public func == <T: Object>(lhs: T, rhs: T) -> Bool {
...
+    public override subscript(key: String) -> AnyObject? {
+        get {
+            if let prop = RLMObjectBaseObjectSchema(self)[key] {
+                if prop.type == RLMPropertyType.Array {
```
Comment (jpsim) [inline RealmSwift/Object.swift]: `prop.type == .Array`

File: RealmSwift/Object.swift [hunk 3]
```diff
@@ -206,7 +216,40 @@ public func == <T: Object>(lhs: T, rhs: T) -> Bool {
...
+                    listProperties[key] = list
+                    return list
+
+                }
+            }
+            return RLMObjectBaseObjectForKeyedSubscript(self, key)
+        }
+        set(value) {
+            if let prop = RLMObjectBaseObjectSchema(self)[key] {
+                if prop.type == RLMPropertyType.Array {
```
Comment (jpsim) [inline RealmSwift/Object.swift]: `prop.type == .Array`

File: RealmSwift/Object.swift [hunk 4]
```diff
@@ -206,7 +216,40 @@ public func == <T: Object>(lhs: T, rhs: T) -> Bool {
...
+                    throwRealmException("Setting List properties is unsupported. Instead you can remove objects from the current List.")
```
Comment (jpsim) [inline RealmSwift/Object.swift]: > Instead, you can add or remove objects

File: RealmSwift/Object.swift [hunk 5]
```diff
@@ -244,4 +287,11 @@ public class ObjectUtil : NSObject {
 
         return properties
     }
+
+    @objc private class func initializeListProperty(object: RLMObjectBase?, property: RLMProperty?, array: RLMArray?) {
+        if let object = object as? Object {
+            let list = object[property!.name]! as RLMListBase
```
Comment (jpsim) [inline RealmSwift/Object.swift]: There are 3 operations in this line that could fail (plus the optional binding above). Throwing an exception if any of these fail will help users (or us) figure out what might have gone wrong.

**Comment by jpsim** [comment]
:+1:

File: Realm/RLMObjectStore.mm [hunk 6]
```diff
@@ -332,20 +332,9 @@ static inline void RLMInitializeSwiftListAccessor(RLMObjectBase *object) {
...
-                [(id<RLMSwiftMigrationObject>)object initalizeListPropertyWithName:prop.name rlmArray:array];
-            }
-            else {
-                auto list = static_cast<RLMListBase *>(object_getIvar(object, prop.swiftListIvar));
-                list._rlmArray = [RLMArrayLinkView arrayWithObjectClassName:prop.objectClassName
-                                                                       view:object->_row.get_linklist(prop.column)
-                                                                      realm:object->_realm];;
-            }
+            [RLMObjectUtilClass(YES) initializeListProperty:object property:prop array:[RLMArrayLinkView arrayWithObjectClassName:prop.objectClassName
+                                                                                                                             view:object->_row.get_linklist(prop.column)
```
Comment (segiddins) [inline Realm/RLMObjectStore.mm]: Why these blank lines?

File: RealmSwift/Migration.swift [hunk 7]
```diff
@@ -29,6 +29,9 @@ Migration block used to migrate a Realm.
 */
 public typealias MigrationBlock = (migration: Migration, oldSchemaVersion: UInt) -> Void
 
+/// Object class used during migrations
+public typealias MigrationObject=DynamicObject
```
Comment (segiddins) [inline RealmSwift/Migration.swift]: space around the `=`

File: RealmSwift/Tests/ObjectTests.swift [hunk 8]
```diff
@@ -170,4 +170,8 @@ class ObjectTests: TestCase {
             test(persistedObject)
         }
     }
+
+    func testSubscript() {
+
```
Comment (segiddins) [inline RealmSwift/Tests/ObjectTests.swift]: add a `TODO`?

Comment (alazier) [inline RealmSwift/Tests/ObjectTests.swift]: There is a github issue

**Pull request merged**